New Workflow Management Screen UI using JavaScript

anuprulez opened the pull request:
# Introduction:
The Pull Request presents a new user interface for workflow's home screen built using JavaScript and thereby retires its mako file (list.mako). It displays list of user's workflows in the center panel of the Galaxy as opposed to previous display in full page. 

# Visuals:

### List of workflows
![workflow_1](https://cloud.githubusercontent.com/assets/3022518/25864618/70a58038-34f0-11e7-9971-43b18f5b124d.png)

### Search
![workflow_2](https://cloud.githubusercontent.com/assets/3022518/25864617/70a48b74-34f0-11e7-9a90-1303b1bed5e4.png)

# Details:
The workflow's table merges the shared and user's own workflows and can be identified by text in the 'owner' column ('You' for user's own workflows and an email id for the shared workflows). Moreover, as an enhancement, it includes a client side search which filters the list of workflows using query typed in the search textbox (query needs at least 3 characters to search).

# Routing:
It uses client side routing in the same way as the user preferences and custom builds screens.

# Credits:
Thanks a lot to @bgruening and @guerler for suggestions and reviews!

Your suggestions are welcome!

anuprulez committed:
Initial commit for transforming mako files for workflow into plain javascript

anuprulez committed:
Create new workflow

anuprulez committed:
Show workflow information using workflow select list

anuprulez committed:
Fixing the routing issue and showing the main workflow dashboard in the center panel

anuprulez committed:
Adding actions dropdown for workflows

anuprulez committed:
Redirecting to workflow list page once actions are over

anuprulez committed:
Adding shared workflow and congifure workflow options

anuprulez committed:
Adding actions for shared workflows

anuprulez committed:
Changing html structure a bit

anuprulez committed:
Updating html and js structure and adding font awesome icons

anuprulez committed:
Adding view action for shared workflow and fixing font-family for buttons

anuprulez committed:
Merging with dev

anuprulez committed:
Adding table for showing all workflows

anuprulez committed:
Adding client side search for workflows

anuprulez committed:
Converting run mako to plain javascript and modifying menu for url

anuprulez committed:
Merging with dev

anuprulez committed:
Moving the action buttons text to their tooltips

anuprulez committed:
Merge branch 'dev' of https://github.com/bgruening/galaxy into workflowexport_mako

anuprulez committed:
Deleting list.mako file and updating code for workflow run

nsoranzo commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -65,6 +65,8 @@ def index(self, trans, **kwd):
             item['url'] = url_for('workflow', id=encoded_id)
             item['owner'] = wf.user.username
+            item['latest_workflow_steps'] = len( wf.latest_workflow.steps )
+            item['published'] = wf.published
```
This line is not necessary, `'published'` field is already filled by the `to_dict` method.

nsoranzo commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
             item['url'] = url_for( 'workflow', id=encoded_id )
             item['owner'] = wf_sa.stored_workflow.user.username
+            item['latest_workflow_steps'] = len( wf_sa.stored_workflow.latest_workflow.steps )
+            item['user_email'] = wf_sa.stored_workflow.user.email
```
I think this line should be removed, it's not necessary to expose the user email through the API, the owner should be enough for what you are doing.

nsoranzo commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
+            item['user_email'] = wf_sa.stored_workflow.user.email
+            item['user_name'] = wf_sa.stored_workflow.user.username
+            item['slug'] = wf_sa.stored_workflow.slug
+            item['published'] = wf_sa.stored_workflow.published            
```
As above, this is not necessary, `'published'` field is already filled by the `to_dict` method.

nsoranzo commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -0,0 +1,223 @@
+                wf_obj.text = wf.name;
+                wf_obj.workflow_steps = wf.latest_workflow_steps;
+                wf_obj.published = ( wf.published ? "Yes" : "No" );
+                wf_obj.email = wf.user_email ? wf.user_email : "You";
```
Remove

nsoranzo commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -0,0 +1,223 @@
+                wf_obj.published = ( wf.published ? "Yes" : "No" );
+                wf_obj.email = wf.user_email ? wf.user_email : "You";
+                wf_obj.slug = wf.slug ? wf.slug : "";
+                wf_obj.username = wf.user_name ? wf.user_name : "";
```
```javascript
wf_obj.username = wf.owner === username ? "You" : wf.owner;
```

nsoranzo commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -74,6 +76,11 @@ def index(self, trans, **kwd):
             item['owner'] = wf_sa.stored_workflow.user.username
+            item['latest_workflow_steps'] = len( wf_sa.stored_workflow.latest_workflow.steps )
+            item['user_email'] = wf_sa.stored_workflow.user.email
+            item['user_name'] = wf_sa.stored_workflow.user.username
```
That's the same as the owner, remove.

anuprulez committed:
Fixing pylint comments for build failure

nsoranzo commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -0,0 +1,223 @@
+                                     self._templateActions( wf ) +
+                                 '</div>' +
+                              '</td>' +
+                              '<td>' + _.escape( wf.email ) +'</td>' +
```
s/email/username/

nsoranzo reviewed (commented):
Another mako gone, thanks a lot @anuprulez!

bgruening commented:
@nsoranzo @erasche if you have time can you test this. The UI changes a little bit and we would like to have some more feedback on this side as well! And thanks for the review Nicola!

hexylena commented:
yep, will do! This looks amazing

anuprulez committed:
Fixing PR review comments

mvdbeek commented:
I just tested this out, I think this is a very nice improvement. The fact that this is rendered in the center panel is great, everything works, I'm impressed.

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -65,6 +65,7 @@ def index(self, trans, **kwd):
             encoded_id = trans.security.encode_id(wf.id)
             item['url'] = url_for('workflow', id=encoded_id)
             item['owner'] = wf.user.username
+            item['latest_workflow_steps'] = len( wf.latest_workflow.steps )
```
`latest_workflow_steps` was a little confusing at first, maybe `step_number` or `number_of_steps` would be clearer?

nsoranzo replied:
Agreed, something like `number_of_steps` would be better.

anuprulez commented:
@dannon When I take latest from dev, "make client" is producing this error:

```
Warning: Uglification failed.
Unexpected token punc «)», expected punc «,». 
Line 328 in ../static/src/nls/de/locale.js
```
I committed the PR review comments but the build failed. So, I reverted the commits to the successful build and again took the latest changes from dev (that include localization changes) but it again shows the same error. Can you help? The locale.js files are present for different languages.

Thanks a lot!

dannon commented:
@anuprulez Yep, taking a look and I'll get a fix out in a sec if it's busted.

(yep, found two problems, fixing them now.  once the client build is a part of the PR merge process instead of manual action, this will be easier to catch)

dannon commented:
@anuprulez Fixed with #4050

anuprulez commented:
@dannon thanks. It seems like a job has failed in the Travis build.

The [build](https://travis-ci.org/galaxyproject/galaxy/jobs/230760116) for [#4050](https://github.com/galaxyproject/galaxy/pull/4050) has failed for this error:
```
>> Masthead test - tabs
>> Message: beforeEach failed on tabs: Can't find variable: l
```
Travis build shows - Job #9001.4 failed

dannon commented:
Yep, I was looking into that, should have a fix out in a sec.

anuprulez committed:
Merging the static files

hexylena commented:
@anuprulez 

Overall:

- Love being able to quickly filter without needing to ctrl-F
- the buttons are clear
- More information in the workflow menu is really nice, it was lacking before.
- Seeing the "workflows shared with you" and "your workflows" mixed together is really nice. It goes part way toward fixing the bugs / features that still bother me in https://github.com/galaxyproject/galaxy/issues/298) 

I did find some minor bugs. Please consider testing with galaxy under a proxy_prefix (e.g. /galaxy/) as it helps detect some of these bugs.

- The three main buttons all do not respect proxy prefix, and lead to 404s (configure your workflow menu / create / upload, all go to `host/workflow/...` rather than `host/galaxy/workflow/...`)
- All options in the dropdown need to use the prefix as well

anuprulez committed:
Merging with dev and resolving conflicts

anuprulez committed:
Fixing PR review comments

anuprulez commented:
@erasche @mvdbeek @nsoranzo @dannon Thanks a lot for your reviews and help!

guerler commented:
How about displaying this also when users click on 'All workflows' in the Tool panel?

anuprulez commented:
@erasche Right now, I am not really sure about the Galaxy's proxy prefix how it works. I will try to explore and work on it. Thanks :)

anuprulez commented:
@guerler we can do this. It will help us remove one more mako (workflow/list_for_run.mako) provided that the list of workflows displayed by /workflow and /workflow/list_for_run (All Workflows in ToolPanel) are same.

anuprulez committed:
Fixing review comments for proxy_prefix and 'All workflows' link in the tool panel

anuprulez commented:
@erasche For proxy_prefix bug, I have updated all the links to contain 'Galaxy.root' before them. But I cannot test them. Does it need proxy server? Locally, they are working fine with updated url.

@guerler I updated the 'All Workflows' link in the ToolPanel to show the new workflows list. I did not remove the "list_for_run.mako" and the python code for it. I looked for "list_for_run" in the entire code repository of Galaxy, I found only two occurrences:

- client/galaxy/scripts/apps/tool-panel.js (already updated)
- lib/galaxy/webapps/galaxy/controllers/workflow.py

So, shall we remove the "list_for_run" python method in the workflow controller and the "list_for_run.mako"?

Thanks!

hexylena commented:
> For proxy_prefix bug, I have updated all the links to contain 'Galaxy.root' before them. But I cannot test them. Does it need proxy server? Locally, they are working fine with updated url.

For future reference, @anuprulez it does not need a proxy server, it isn't truly for proxying, just for "run under this URL sub-path". The following galaxy.ini config changes are sufficient to test, and then you can access your galaxy at /galaxy

```
[filter:proxy-prefix]                                                                                                                              
use = egg:PasteDeploy#prefix                                                                                                                       
prefix = /galaxy                                                                                                                                   
[app:main]
filter-with = proxy-prefix                                                              
cookie_path = /galaxy                                                                                                                              
```

That said, I have tested it, it looks good now! Small visual artefact. This looks correct:

![utvalg_098](https://cloud.githubusercontent.com/assets/458683/25925084/d17cc3ae-35d5-11e7-9122-b85eff398e87.png)

But this seems to have no spacing around it:

![utvalg_099](https://cloud.githubusercontent.com/assets/458683/25925083/d1773fce-35d5-11e7-9674-318b231a750f.png)

anuprulez committed:
Fixing for prefix url for actions in workflow controller

guerler commented:
@anuprulez Yeah, I think we should remove it.

anuprulez commented:
@erasche thanks a lot :)

Using those configuration, i could have "prefix" in the url and tested for all the actions like edit, run etc. There were couple of issues with the url again while redirecting the page from server. I have fixed those. Now all actions work fine (with and without prefix in the url).

The visuals you shared come from another mako file (import or upload workflow) and form (create workflow) from server which we have not worked upon in this PR. In future, we plan to transform those mako files as well.

Thanks a lot for your review and help!

anuprulez committed:
Removing list_for_run.mako and its corresponding method in workflow controller

anuprulez commented:
@guerler I removed "list_for_run.mako" file and its corresponding method in the workflow controller.

bgruening commented:
It's green! thanks all for your very nice comments!
Anyone wants to merge this! More to come soon :)

jmchilton committed:
Fix Selenium test cases for JavaScript-izing workflow management.

jmchilton commented:
Really fantastic contribution and nice work! I've opened a PR with Selenium test fixes for these changes (https://github.com/bgruening/galaxy/pull/34).

bgruening committed:
Merge pull request #34 from jmchilton/workflowexport_mako_test_fixes

Fix Selenium test cases for JavaScript-izing workflow management.

nsoranzo commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -0,0 +1,201 @@
+                else {
+                    $el_workflow.append( self._templateNoWorkflow() );
+                }
+            });
```
Indentation of lines 18-40 seems wrong.

dannon replied:
No, this is correct, unless I misread your concern.   The following is a single statement, not the opening of a block:
```
var self = this,
    min_query_length = 3;
```

Personally I think it's less clear than explicit var declarations, but multiple assignment like this is fine.

anuprulez replied:
The method `render` starts at column 9 
`$.getJSON ` starts at column 13 (line 18) and ends at column 13 (line 40).
Inside `$.getJSON `, lines start at column 17 and inside `if-else` lines start at column 21

nsoranzo replied:
Oh how I hate this language! Sorry for the noise!

dannon committed:
Fix require inclusion scoping in analysis.js

dannon commented on client/galaxy/scripts/apps/analysis.js:
```diff
@@ -11,6 +11,7 @@ var jQuery = require( 'jquery' ),
     UserPreferences = require( 'mvc/user/user-preferences' );
     CustomBuilds = require( 'mvc/user/user-custom-builds' );
     Tours = require( 'mvc/tours' );
+    Workflows = require( 'mvc/workflow/workflow' );
```
This is made global due to the trailing semicolons above.  Fixed in https://github.com/bgruening/galaxy/pull/35/.

dannon commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -0,0 +1,201 @@
+define( [ 'mvc/form/form-view',
+          'mvc/ui/ui-misc',
+          'mvc/ui/ui-select',
+          'mvc/tool/tool-form-composite' ], function( Form, Ui, Select, ToolForm ) {
```
It doesn't look like we're actually using any of these here, or am I missing it?

anuprulez replied:
I am looking into it. Thanks!

dannon replied:
Sure!  Pending that, I'm happy to merge this.  Don't worry about rebuilding the client again if there are no additional changes, I can resolve that in the merge.

anuprulez replied:
Actually I rebuilt the client after removing the imports from the header of workflow.js to verify they are not needed and now I have all those static files modified. Shall I commit all these static files along with conflicting files?

### This branch has conflicts that must be resolved
```
Conflicting files
static/scripts/bundled/libs.bundled.js
static/scripts/bundled/libs.bundled.js.map
```

dannon replied:
That'd be fine, sure.

anuprulez replied:
I pushed the changes. The build is running :)

dannon replied:
Ahh, k, I see.  That last merge with dev wasn't necessary, but I can wait on the build and re-merge after if you want.

anuprulez replied:
Ok, sure. Thank you!

bgruening committed:
Merge pull request #35 from dannon/workflowexport_mako

Fix require inclusion scoping in analysis.js

anuprulez committed:
Removing unused imports in workflow main view

anuprulez committed:
Merging with dev

dannon commented:
@galaxybot test this

dannon commented:
(probably should have asked Travis to test this before, sorry!)

dannon commented:
Thanks @anuprulez, nice work!

anuprulez commented:
Thanks a lot everyone for comments and help!
